### ISSUE TITLE
Direct data access 'filter' parameter

### ISSUE THREAD

**Issue by sambedingfield** [directus/docs#181]
Loving Directus, but I've been halted by most likely a simple syntax thing.

I'm following the docs for [TableGateway Factory](https://docs.directus.io/advanced/api/data.html#fetching-data-within-hooks-and-endpoints) as I'm trying to access direct data within a custom API endpoint. However, I'm struggling with the 'filter' parameter, everything I try results in an _Internal Server Error_.

For instance, the standard API might be: 
`?single=1&filter[id][eq]=1&filter[name][eq]=sam?`

How would this be expected within the TableGateway parameter?

```
$item = $tableGateway->getItems([
    'single' => true,
    'filter' => [
        // how should the key, value and operator be formatted?
    ]
]);
````

I know I could use Zend, but I'd prefer to use the Directus class.

Also, my apologies if this is in the wrong place, I wasn't sure which repo to post under.

**Comment by Lapsus** [comment]
this is default PHP, as it translates one-to-one. 
URL parameters are converted to a PHP array and passed as parameter.
Your example URL would look like this for example. 
```
$item = $tableGateway->getItems([
    'single' => true,
    'filter' => [
        'id' => [
             'eq' => 1
        ],
       'name' => [
           'eq' => 'sam?'
       ]
    ]
]);
```

**Comment by sambedingfield** [comment]
Well that's what I would have thought, but the API then results in an error.

Here's the example custom endpoint I'm using:
```
use Directus\Application\Http\Request;
use Directus\Application\Http\Response;

return [
    '' => [
        'method' => 'GET',
        'handler' => function (Request $request, Response $response) {
            $container = \Directus\Application\Application::getInstance()->getContainer();
            $dbConnection = $container->get('database');
            $tableGateway = Directus\Database\TableGatewayFactory::create('codes', [
                'connection' => $dbConnection,
                'acl' => false
            ]);
            $item = $tableGateway->getItems([
                'single' => true,
                'filter' => []
            ]);
            return $response->withJson($item);
        }
    ]
];
```
The above works correctly and the API returns the expected first result of the selected "codes" table.

But when I change the params to:
```
$item = $tableGateway->getItems([
    'single' => true,
    'filter' => [
        'id' => [
            'eq' => 1
        ]
    ]
]);
```
The API responds with `{"error":{"code":null,"message":"Internal Server Error"}}`

**Comment by Lapsus** [comment]
Could you please check the logs for the error message? That might help a lot actually.

**Comment by sambedingfield** [comment]
```
[2019-07-31 09:32:59] api[_].ERROR: Error: Call to a member function getStatusesOnReadFieldBlacklist() on null in \src\core\Directus\Database\TableGateway\RelationalTableGateway.php:1618
Stack trace:
#0 [internal function]: Directus\Database\TableGateway\RelationalTableGateway->processFilter
#1 \src\core\Directus\Database\TableGateway\RelationalTableGateway.php(819): call_user_func_array
#2 \src\core\Directus\Database\TableGateway\RelationalTableGateway.php(1120): Directus\Database\TableGateway\RelationalTableGateway->applyParamsToTableEntriesSelect
#3 \src\core\Directus\Database\TableGateway\RelationalTableGateway.php(1248): Directus\Database\TableGateway\RelationalTableGateway->fetchItems
#4 \src\core\Directus\Database\TableGateway\RelationalTableGateway.php(844): Directus\Database\TableGateway\RelationalTableGateway->fetchData
#5 \public\extensions\custom\endpoints\code.php(21): Directus\Database\TableGateway\RelationalTableGateway->getItems
#6 [internal function]: Closure->{closure}
#7 \vendor\slim\slim\Slim\Handlers\Strategies\RequestResponse.php(41): call_user_func
#8 \vendor\slim\slim\Slim\Route.php(356): Slim\Handlers\Strategies\RequestResponse->__invoke
#9 \src\core\Directus\Application\Http\Middleware\AuthenticationMiddleware.php(109): Slim\Route->__invoke
#10 [internal function]: Directus\Application\Http\Middleware\AuthenticationMiddleware->__invoke
#11 \vendor\slim\slim\Slim\DeferredCallable.php(43): call_user_func_array
#12 [internal function]: Slim\DeferredCallable->__invoke
#13 \vendor\slim\slim\Slim\MiddlewareAwareTrait.php(70): call_user_func
#14 \src\core\Directus\Application\Http\Middleware\TableGatewayMiddleware.php(25): Slim\Route->Slim\{closure}
#15 [internal function]: Directus\Application\Http\Middleware\TableGatewayMiddleware->__invoke
#16 \vendor\slim\slim\Slim\DeferredCallable.php(43): call_user_func_array
#17 [internal function]: Slim\DeferredCallable->__invoke
#18 \vendor\slim\slim\Slim\MiddlewareAwareTrait.php(70): call_user_func
#19 \vendor\slim\slim\Slim\MiddlewareAwareTrait.php(117): Slim\Route->Slim\{closure}
#20 \vendor\slim\slim\Slim\Route.php(334): Slim\Route->callMiddlewareStack
#21 \vendor\slim\slim\Slim\App.php(515): Slim\Route->run
#22 \src\core\Directus\Application\Http\Middleware\AbstractRateLimitMiddleware.php(34): Slim\App->__invoke
#23 [internal function]: Directus\Application\Http\Middleware\AbstractRateLimitMiddleware->__invoke
#24 \vendor\slim\slim\Slim\DeferredCallable.php(43): call_user_func_array
#25 [internal function]: Slim\DeferredCallable->__invoke
#26 \vendor\slim\slim\Slim\MiddlewareAwareTrait.php(70): call_user_func
#27 \vendor\directus\proxy-detection\src\ProxyDetectionMiddleware.php(30): Slim\App->Slim\{closure}
#28 \src\core\Directus\Application\Http\Middleware\ProxyMiddleware.php(18): RKA\Middleware\ProxyDetectionMiddleware->__invoke
#29 [internal function]: Directus\Application\Http\Middleware\ProxyMiddleware->__invoke
#30 \vendor\slim\slim\Slim\DeferredCallable.php(43): call_user_func_array
#31 [internal function]: Slim\DeferredCallable->__invoke
#32 \vendor\slim\slim\Slim\MiddlewareAwareTrait.php(70): call_user_func
#33 \vendor\akrabat\ip-address-middleware\src\IpAddress.php(113): Slim\App->Slim\{closure}
#34 [internal function]: RKA\Middleware\IpAddress->__invoke
#35 \vendor\slim\slim\Slim\DeferredCallable.php(43): call_user_func_array
#36 [internal function]: Slim\DeferredCallable->__invoke
#37 \vendor\slim\slim\Slim\MiddlewareAwareTrait.php(70): call_user_func
#38 \src\core\Directus\Application\Http\Middleware\CorsMiddleware.php(66): Slim\App->Slim\{closure}
#39 [internal function]: Directus\Application\Http\Middleware\CorsMiddleware->__invoke
#40 \vendor\slim\slim\Slim\DeferredCallable.php(43): call_user_func_array
#41 [internal function]: Slim\DeferredCallable->__invoke
#42 \vendor\slim\slim\Slim\MiddlewareAwareTrait.php(70): call_user_func
#43 \src\core\Directus\Application\Http\Middleware\ResponseCacheMiddleware.php(54): Slim\App->Slim\{closure}
#44 [internal function]: Directus\Application\Http\Middleware\ResponseCacheMiddleware->__invoke
#45 \vendor\slim\slim\Slim\DeferredCallable.php(43): call_user_func_array
#46 [internal function]: Slim\DeferredCallable->__invoke
#47 \vendor\slim\slim\Slim\MiddlewareAwareTrait.php(70): call_user_func
#48 \vendor\slim\slim\Slim\MiddlewareAwareTrait.php(117): Slim\App->Slim\{closure}
#49 \vendor\slim\slim\Slim\App.php(405): Slim\App->callMiddlewareStack
#50 \vendor\slim\slim\Slim\App.php(313): Slim\App->process
#51 \src\core\Directus\Application\Application.php(161): Slim\App->run
#52 \public\index.php(5): Directus\Application\Application->run [] []
```

Sorry, I'm new to using Directus! I'm using the latest Directus Suite from a couple days ago and followed all docs until this far.

**Comment by Lapsus** [comment]
when creating your tableGateway you pass along acl = false.
But the tableGateway checks for permissions when parsing the filter. As these are not present, it fails. Please remove acl false from the tableGateway factory call and try again.

**Comment by sambedingfield** [comment]
I see.

What I'm trying to do is create a public API endpoint, which checks for an item within a private collection. Is this not the best way of achieving this? And could you point me in the right direction (any docs I've missed)?

Thanks

**Comment by Lapsus** [comment]
I think the docs are fine. But it looks strange to me what you want to do here.

You basically want to circumvent permissions and acl, but why?
If the collection is private it is private. No matter how you try to access it. (guess you could use the database directly but why use directus than at all.

If you want to read the data (public), but prevent create/edit/delete, you could grant public read permissions to your collection and create/edit/delete to a user role.

You do not need a custom endpoint then to fetch the data. Just use the ItemsEndpoint that already exists.

Also i suggest you close the issue. If you have questions like this use Slack. This is not the place for discussions like this in my opinion.

**Comment by sambedingfield** [comment]
No problem with the docs, I just wondered if I'd interpreted something incorrectly or missed something.

It was the type of error that threw me off (not being permission related) and I presumed that all database access should be ideally done through Directus (for all reading and writing) when creating custom endpoints.

I tried Slack but it wasn't much help - but I'll continue discussion there. Sorry!
